reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,129 @@
 
 package kotlinx.coroutines.rx2
 
-import io.reactivex.Scheduler
+import io.reactivex.*
+import io.reactivex.disposables.*
+import io.reactivex.plugins.*
 import kotlinx.coroutines.*
-import java.util.concurrent.TimeUnit
-import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.channels.*
+import java.util.concurrent.*
+import kotlin.coroutines.*
 
 /**
  * Converts an instance of [Scheduler] to an implementation of [CoroutineDispatcher]
  * and provides native support of [delay] and [withTimeout].
  */
-public fun Scheduler.asCoroutineDispatcher(): SchedulerCoroutineDispatcher = SchedulerCoroutineDispatcher(this)
+public fun Scheduler.asCoroutineDispatcher(): CoroutineDispatcher =
+    if (this is DispatcherScheduler) {
+        dispatcher
+    } else {
+        SchedulerCoroutineDispatcher(this)
+    }
+
+/**
+ * Converts an instance of [CoroutineDispatcher] to an implementation of [Scheduler].
+ */
+public fun CoroutineDispatcher.asScheduler(): Scheduler =
+    if (this is SchedulerCoroutineDispatcher) {
+        scheduler
+    } else {
+        DispatcherScheduler(this)
+    }
+
+private class DispatcherScheduler(internal val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    private val job = SupervisorJob()
+    private val scope = CoroutineScope(job + dispatcher)
+
+    override fun scheduleDirect(block: Runnable): Disposable {
+        if (scope.isActive) {
+            return scope.launch {
+                block.runWithRx()
+            }.asDisposable()
+        }
+
+        return Disposables.disposed()
+    }
+
+    override fun scheduleDirect(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        if (delay <= 0) {
+            return scheduleDirect(block)
+        }
+
+        if (scope.isActive) {
+            return scope.launch {
+                delay(unit.toMillis(delay))
+                block.runWithRx()
+            }.asDisposable()
+        }
+
+        return Disposables.disposed()
+    }
+
+    private fun Runnable.runWithRx() = RxJavaPlugins.onSchedule(this).run()
```

Comment:
@zach-klippenstein from what I'm reading in the docs, the reason they use ```SchedulerRunnableIntrospection``` is because the ```Scheduler.schedule``` calls ```Scheduler.Worker.schedule```. This leads to multiple ```RxJavaPlugins.onSchedule(block)``` calls since you're supposed to call that method in both classes. 

However, since I'm not delegating to the worker in the scheduler in ```Scheduler.schedule```, I don't believe I need to use ``SchedulerRunnableIntrospection```. 

Let me know if that makes sense.

Reply:
docs don't say explicitly that custom classes should use it.

Reply:
You're right, I misread the docs.